include/xtypes/MapInstance.hpp

```diff
@@ -0,0 +1,416 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef EPROSIMA_XTYPES_MAP_INSTANCE_HPP_
+#define EPROSIMA_XTYPES_MAP_INSTANCE_HPP_
+
+#include <xtypes/PairType.hpp>
+
+#include <cstdint>
+#include <cstring>
+
+namespace eprosima {
+namespace xtypes {
+
+/// \brief Implementation of a dynamic map of DynamicTypes.
+/// This class is used internal by MapType to implement its behaviour.
+class MapInstance
+{
+public:
+    /// \brief Construct a MapInstance
+    /// \param[in] content Content of the map
+    /// \param[in] capacity Reserved memory for the map.
+    MapInstance(
+            const PairType& content,
+            uint32_t capacity = 0)
+        : content_(content)
+        , block_size_(content.memory_size())
+        , capacity_(capacity)
+        , memory_(capacity > 0 ? new uint8_t[capacity * block_size_] : nullptr)
+        , size_(0)
+    {}
+
+    MapInstance(
+            const MapInstance& other)
+        : content_(other.content_)
+        , block_size_(other.block_size_)
+        , capacity_(other.capacity_)
+        , memory_(capacity_ > 0 ? new uint8_t[capacity_ * block_size_] : nullptr)
+        , size_(other.size_)
+    {
+        if(memory_ != nullptr)
+        {
+            copy_content(memory_, other.memory_);
+        }
+    }
+
+    /// \brief Copy constructor from a MapInstance with other but compatible content.
+    /// \param[in] other Map from copy the values.
+    /// \param[in] content Content of the map
+    /// \param[in] bounds Max copied elements.
+    /// \pre content and other.content compatibles (see dds::core::xtypes::DynamicType::is_compatible())
+    MapInstance(
+            const MapInstance& other,
+            const PairType& content,
+            uint32_t bounds)
+        : content_(content)
+        , block_size_(content.memory_size())
+        , capacity_(bounds == 0 ? other.capacity_ : std::min(other.capacity_, bounds))
+        , memory_(capacity_ > 0 ? new uint8_t[capacity_ * block_size_] : nullptr)
+        , size_(bounds == 0 ? other.size_ : std::min(other.size_, bounds))
+    {
+        if(memory_ != nullptr)
+        {
+            copy_content_from_type(memory_, other.memory_, other.content_);
+        }
+    }
+
+    MapInstance(
+            MapInstance&& other)
+        : content_(std::move(other.content_))
+        , block_size_(std::move(other.block_size_))
+        , capacity_(std::move(other.capacity_))
+        , memory_(std::move(other.memory_))
+        , size_(std::move(other.size_))
+    {
+        other.memory_ = nullptr;
+    }
+
+    /// \brief Deep equality operator
+    bool operator == (
+            const MapInstance& other) const
+    {
+        if(other.size() != size_)
+        {
+            return false;
+        }
+
+        if(content_.first().is_constructed_type() || content_.second().is_constructed_type())
```

Comment:
```suggestion
        if(content_.first().is_constructed_type() && content_.second().is_constructed_type())
```

But check in `else` both are primitives. If not, return error.

Reply:
Not in this case. We are comparing `PairType` instances. We check the components of the pair just to try the optimization.